### PR TITLE
Refunds in Order Details: aggregate data and item sorting

### LINKED ISSUE #1717: Refunded Products: algorithm to combine product data
### As a user, I want to see summarized data in the Refunded Products screen, no matter how many refunds the same product type comes from.

In the Order Details screen and Refunded Products screen, the user expects to see a different data format when a "refund by item" has taken place on their website. The problem is that the refunded order items array is indiscriminate and wants to show refunded order items grouped together by their individual refund transactions (refundID and orderItemID), *not* by their product and variation IDs.

The problem, in pictures:

The user's expectation, in pictures:

This will require:
 
- an algorithm that combines the existing data that the user perceives as "duplicates", into a summarized data format
- a new model
- a new view model
- editing the ProductDetailsTableViewCell to accept the new viewModel

**Do not** take the new data and attempt to shoe-horn it into the OrderItemRefund type. OrderItemRefund type is read-only data that comes from the API and is a struct with constants. The new type will need to allow computed properties and variables (and might need to be a class).

### LINKED ISSUE #1731: Order Details: aggregate data and sorting
If an order has refunded products, draw in the table cells to use the updated information. Make the order items "do the math" from the original reported quantities and totals for an order item and subtract any refunded items.

- [ ] add a sort strategy for Order Details > order items
- [ ] create aggregate data for order items  (order items - refunded products = current qty)
- [ ] display the new aggregate data cells

### PR DESCRIPTION
**Pull request by mindgraffiti**
Closes #1731 and Closes #1717 

Ref. p99K0U-2no-p2

This PR calculates the total quantities remaining for the listed order items and also re-calculates the order item totals based on the new quantity. The order items are sorted by productID and variationID, in ascending order. The Refunded Products screen is also sorted the same way.

### Screenshots
**Before**

**After**

### To test
The goal is to create multiple refunds with several of the same product type(s) in the refunds.

1. On your test site, have an order with multiple product types. Make sure to note the Order Number.
2. Find the order and "refund by item".
3. Make another refund. Refund a couple more of the same product, and add another product to make it interesting.
4. Make another refund. Refund a couple more of the same product(s) or refund a different item by itself.
5. Check out the develop branch and navigate in the app to: Orders tab > Order Number (Order Details screen). 
6. Check out this branch. Navigate to the same location. The cells should now be summarized.

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by mindgraffiti** [comment]
Agreed @jaclync, this is really hard to follow. `OrderItemRefundSummary` is supposed to be the aggregator struct. Maybe I need to rename it and then add a second one that only represents Refunded Products. I need to think on it rather than keep throwing code in it. 🤔

File: WooCommerce/Classes/ViewModels/Order Details/Aggregate Order Items/AggregateOrderItemCellViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,92 @@
+import Foundation
+import Yosemite
+
+
+// MARK: - View Model for an aggregate order item cell
+//
+struct AggregateOrderItemCellViewModel {
```
Comment (jaclync) [inline WooCommerce/Classes/ViewModels/Order Details/Aggregate Order Items/AggregateOrderItemCellViewModel.swift]: from the diffs in `ProductDetailsTableViewCell`, it looks like this struct only has to expose `imageURL`, `name`, `quantity`, `price`, and `sku`. could you make sure the other properties are private?

Comment (jaclync) [reply WooCommerce/Classes/ViewModels/Order Details/Aggregate Order Items/AggregateOrderItemCellViewModel.swift]: after taking a pass of the code, these 3 view models feel like rendering the same fields:
- `AggregateOrderItemCellViewModel`
- `RefundedOrderItemCellViewModel` 
- `RefundedProductCellViewModel`

even though these three view models derive the view model fields from different data models, if the fields are the same (look like they all expose `imageURL`, `name`, `quantity`, `price`, and `sku`), I'd suggest making one cell view model and implement different initializers for each data model. Otherwise, if we ever update the price formatting or SKU strings, we'd have to remember updating all 3 places.

I might have missed some other substantial differences between these 3 view models, and feel free to ignore this comment if that's the case!

File: WooCommerce/Classes/ViewModels/Order Details/Refunded Products/RefundedProductsDataSource.swift [hunk 2]
```diff
@@ -139,7 +140,7 @@ extension RefundedProductsDataSource {
     ///
     func reloadSections() {
         let refundedProducts: Section? = {
-            let rows: [Row] = Array(repeating: .orderItemRefunded, count: items.count)
+            let rows: [Row] = Array(repeating: .orderItemRefunded, count: self.refundedProducts.count)
```
Comment (jaclync) [inline WooCommerce/Classes/ViewModels/Order Details/Refunded Products/RefundedProductsDataSource.swift]: nit: `self` might not be necessary?

File: WooCommerce/Classes/ViewModels/Order Details/Refund Details/RefundedOrderItemCellViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,108 @@
+import Foundation
+import Yosemite
+
+
+// MARK: - View Model for a refunded order item cell
+//
+struct RefundedOrderItemCellViewModel {
```
Comment (jaclync) [inline WooCommerce/Classes/ViewModels/Order Details/Refund Details/RefundedOrderItemCellViewModel.swift]: similar to the previous comment, it looks like this struct only has to expose `imageURL`, `name`, `quantity`, `price`, and `sku`. could you make sure the other properties are private?

File: WooCommerce/Classes/ViewModels/Order Details/Refunded Products/RefundedProduct.swift [hunk 4]
```diff
@@ -0,0 +1,64 @@
+import Foundation
+
+
+/// This model represents the aggregate data for refunded products.
+///
+final class RefundedProduct {
```
Comment (jaclync) [inline WooCommerce/Classes/ViewModels/Order Details/Refunded Products/RefundedProduct.swift]: nit: could this be a struct?

Comment (jaclync) [inline WooCommerce/Classes/ViewModels/Order Details/Refunded Products/RefundedProduct.swift]: I'm a bit confused, this model looks like exactly the same as `AggregateOrderItem`, could these two use cases share the same model if the difference is only how the properties populated?

File: WooCommerce/Classes/ViewModels/Order Details/Refunded Products/RefundedProductCellViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,95 @@
+import Foundation
+import Yosemite
+
+
+// MARK: - View Model for a refunded product cell
+//
+struct RefundedProductCellViewModel {
```
Comment (jaclync) [inline WooCommerce/Classes/ViewModels/Order Details/Refunded Products/RefundedProductCellViewModel.swift]: similar to the previous comment, it looks like this struct only has to expose `imageURL`, `name`, `quantity`, `price`, and `sku`. could you make sure the other properties are private?

File: WooCommerce/Classes/ViewModels/Order Details/Refund Details/RefundedOrderItemCellViewModel.swift [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+                                              comment: "<item total> (<item individual price> multipled by <quantity>)")
+        let priceText = String.localizedStringWithFormat(priceTemplate, itemTotal, itemSubtotal, quantity)
+
+        return priceText
+    }
+
+    /// Item's Tax
+    /// Return $0.00 if there is no tax.
+    ///
+    var tax: String? {
```
Comment (jaclync) [inline WooCommerce/Classes/ViewModels/Order Details/Refund Details/RefundedOrderItemCellViewModel.swift]: nit: is tax used?

File: WooCommerce/Classes/ViewModels/Order Details/OrderDetailsDataSource.swift [hunk 7]
```diff
@@ -111,9 +111,133 @@ final class OrderDetailsDataSource: NSObject {
...
+    /// Calculate the new order item quantities and totals after refunded products have altered the fields
+    ///
+    var aggregateOrderItems: [AggregateOrderItem] {
+        if refundedProducts.count == 0 {
+            fatalError("Error: attemtpted to calculate aggregate order item data with no refunded products.")
+        }
+
+        let currency = CurrencyFormatter()
+        // Convert the order items into a mutable type
+        var convertedItems = [AggregateOrderItem]()
```
Comment (jaclync) [inline WooCommerce/Classes/ViewModels/Order Details/OrderDetailsDataSource.swift]: nit: instead of the for loop, can also use `map` so that it's easier to read and the array can be immutable:

```
let convertedItems = items.map { AggregateOrderItem(...) }
```

File: WooCommerce/Classes/ViewModels/Order Details/OrderDetailsDataSource.swift [hunk 8]
```diff
@@ -111,9 +111,133 @@ final class OrderDetailsDataSource: NSObject {
...
+                price: item.price,
+                quantity: item.quantity,
+                sku: item.sku,
+                total: total
+            )
+            convertedItems.append(convertedItem)
+        }
+
+        // Convert refunded products into the same compatible type
+        var convertedRefundedProducts = [AggregateOrderItem]()
```
Comment (jaclync) [inline WooCommerce/Classes/ViewModels/Order Details/OrderDetailsDataSource.swift]: similar to the above comment, can also use `map`:

```
let convertedRefundedProducts = refundedProducts.map { AggregateOrderItem(...) }
```

File: WooCommerce/Classes/ViewModels/Order Details/OrderDetailsDataSource.swift [hunk 9]
```diff
@@ -111,9 +111,133 @@ final class OrderDetailsDataSource: NSObject {
...
+                                                              variationID: refundedProduct.variationID,
+                                                              name: refundedProduct.name,
+                                                              price: refundedProduct.price,
+                                                              quantity: refundedProduct.quantity,
+                                                              sku: refundedProduct.sku,
+                                                              total: refundedProduct.total)
+            convertedRefundedProducts.append(convertedRefundedProduct)
+        }
+
+        var allItems = [AggregateOrderItem]()
```
Comment (jaclync) [inline WooCommerce/Classes/ViewModels/Order Details/OrderDetailsDataSource.swift]: nit: these 3 lines could be:

```
let allItems = convertedItems + convertedRefundedProducts
```

File: WooCommerce/Classes/ViewModels/Order Details/OrderDetailsViewModel.swift [hunk 10]
```diff
@@ -38,6 +38,12 @@ final class OrderDetailsViewModel {
         return dataSource.products
     }
 
+    /// Sorted order items
+    ///
+    var items: [OrderItem] {
```
Comment (jaclync) [inline WooCommerce/Classes/ViewModels/Order Details/OrderDetailsViewModel.swift]: nit: is this used externally? if not, could this be private?

File: WooCommerce/Classes/ViewModels/Order Details/Aggregate Order Items/AggregateOrderItem.swift [hunk 11]
```diff
@@ -0,0 +1,65 @@
+import Foundation
+
+
+/// This model represents a computed summary of order items.
+/// (order items - refunded order items) = aggregate order item data.
+///
+final class AggregateOrderItem {
```
[the remark this answers is not in the document]
Comment (mindgraffiti) [reply WooCommerce/Classes/ViewModels/Order Details/Aggregate Order Items/AggregateOrderItem.swift]: This is a class because I planned to attach a new property in a "helper method" extension. Added it to this PR.

File: WooCommerce/Classes/ViewModels/Order Details/Refunded Products/RefundedProduct.swift (hunk 4 above)
Comment (mindgraffiti) [reply WooCommerce/Classes/ViewModels/Order Details/Refunded Products/RefundedProduct.swift]: Deleting

File: WooCommerce/Classes/ViewModels/Order Details/OrderDetailsDataSource.swift (hunk 7 above)
Comment (mindgraffiti) [reply WooCommerce/Classes/ViewModels/Order Details/OrderDetailsDataSource.swift]: It's still a `var` for convertedItems because map is appending the new types to the array. I like using `map`, though 💯

File: WooCommerce/Classes/ViewModels/Order Details/Refunded Products/RefundedProductsDataSource.swift (hunk 2 above)
Comment (mindgraffiti) [reply WooCommerce/Classes/ViewModels/Order Details/Refunded Products/RefundedProductsDataSource.swift]: Poor naming on my part. The `refundedProducts` section is named the same as the `refundedProducts` class member. Fixed ✅

**Comment by mindgraffiti** [comment]
@jaclync ready for another review. 

Changes include:
- condense 4 `*CellViewModel`s (`OrderItemViewModel`, `AggregateOrderItemCellViewModel`, `RefundedOrderItemCellViewModel`, and `RefundedProductCellViewModel`) into 1, (`ProductDetailsCellViewModel`) and use it everywhere
- add multiple initializers so the new cell view model can be used by multiple data sources
- remove `RefundedProduct` as a type and use `AggregateOrderItem` instead
- instead of calculating `refundedProducts` array twice (once in `OrderDetailsDataSource` and once in `RefundedProductsViewModel`), create a helper class that calculates it, then passes it from the `OrderDetailsViewModel` to the `RefundedProductsViewModel`
- added a convenience property, `productOrVariationID` to check for a valid variationID, instead of performing the check everywhere a productID is needed
- add unit tests for the `AggregateDataHelper` class

File: WooCommerce/Classes/ViewModels/Order Details/Refunded Products/RefundedProductsDataSource.swift [hunk 12]
```diff
@@ -7,9 +7,9 @@ import Yosemite
 /// Must conform to NSObject so it can be the UITableViewDataSource.
 ///
 final class RefundedProductsDataSource: NSObject {
-    /// Refunds
+    /// Aggregate data for refunded products
     ///
-    private var items: [OrderItemRefundSummary]
+    private(set) var refundedProducts: [AggregateOrderItem]
```
Comment (jaclync) [inline WooCommerce/Classes/ViewModels/Order Details/Refunded Products/RefundedProductsDataSource.swift]: nit: could this be `private let` since it doesn't seem to be modified after init?

File: WooCommerce/Classes/ViewModels/Order Details/Aggregate Order Items/AggregateOrderItem.swift (hunk 11 above)
[the remark this answers is not in the document]
Comment (jaclync) [reply WooCommerce/Classes/ViewModels/Order Details/Aggregate Order Items/AggregateOrderItem.swift]: another suggestion: might be worth mentioning "Product" in the name, since it's representing a Product related to an Order

File: WooCommerce/Classes/Tools/AggregateData/AggregateDataHelper.swift [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+    static func refundedProductsCount(from refunds: [Refund]) -> Decimal {
+        var refundedItems = [OrderItemRefund]()
+        for refund in refunds {
+            refundedItems.append(contentsOf: refund.items)
+        }
+
+        var quantities = [Decimal]()
+        for item in refundedItems {
+            quantities.append(item.quantity)
+        }
```
Comment (jaclync) [inline WooCommerce/Classes/Tools/AggregateData/AggregateDataHelper.swift]: nit: the above lines could be simplified into:

```
let refundedItems = refunds.flatMap({ $0.items })
let quantities = refundedItems.map({ $0.quantity })
```

File: WooCommerce/Classes/ViewModels/Order Details/Refunded Products/RefundedProductsViewModel.swift [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+    ///
+    init(order: Order, refundedProducts: [AggregateOrderItem]) {
+        self.order = order
+        self.refundedProducts = refundedProducts
+    }
+
+    /// Update the view model's order when notified
+    ///
+    func update(order newOrder: Order) {
+        self.order = newOrder
```
Comment (jaclync) [inline WooCommerce/Classes/ViewModels/Order Details/Refunded Products/RefundedProductsViewModel.swift]: do you know if other properties need to be updated when the Order is updated?

File: WooCommerce/Classes/ViewModels/Order Details/Aggregate Order Items/AggregateOrderItem.swift (hunk 11 above)
[the remark this answers is not in the document]
Comment (mindgraffiti) [reply WooCommerce/Classes/ViewModels/Order Details/Aggregate Order Items/AggregateOrderItem.swift]: Structs can't have properties in extensions. So I could change this to a struct and add the property to the top. I don't have a good reason other than code organization, since the other things I was experimenting with don't exist anymore 😀

File: WooCommerce/Classes/ViewModels/Order Details/OrderDetailsDataSource.swift [hunk 15]
```diff
@@ -130,7 +130,8 @@ final class OrderDetailsDataSource: NSObject {
         }
     }
 
-    /// Note of customer about the order
+    /// Note from the customer about the order
+    //
```
[the remark this answers is not in the document]
Comment (mindgraffiti) [reply WooCommerce/Classes/ViewModels/Order Details/OrderDetailsDataSource.swift]: 🤦‍♀  I wish Hound could catch those. Thank you!

File: WooCommerce/Classes/ViewModels/Order Details/Refunded Products/RefundedProductsViewModel.swift (hunk 14 above)
Comment (mindgraffiti) [reply WooCommerce/Classes/ViewModels/Order Details/Refunded Products/RefundedProductsViewModel.swift]: 🤔 the OrderStatus is also updated, in `updateOrderStatus(order:)`, and is the function that calls this one. I think that's it.

File: WooCommerce/Classes/ViewModels/ProductDetailsCellViewModel.swift [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+    ///
+    init(item: OrderItem,
+         currency: String,
+         formatter: CurrencyFormatter = CurrencyFormatter(),
+         product: Product? = nil) {
+        self.currency = currency
+        self.currencyFormatter = formatter
+        self.product = product
+        self.name = item.name
+        self.positiveQuantity = abs(item.quantity)
```
[the remark this answers is not in the document]
Comment (mindgraffiti) [reply WooCommerce/Classes/ViewModels/ProductDetailsCellViewModel.swift]: So the API reports all refunds as negative quantities, negative totals, and negative taxes. Because of the context of the screen (a list of all refunded products), it makes sense that the values are shown as positive (as the designs illustrate it). If you check the website, they are negative values because the page displays the math inline. I have preserved the negative values in the data models so that the math is correct when creating the aggregate order items.

File: WooCommerce/Classes/Tools/AggregateData/AggregateDataHelper.swift (hunk 13 above)
Comment (mindgraffiti) [reply WooCommerce/Classes/Tools/AggregateData/AggregateDataHelper.swift]: I'm leaving these changes for my next PR, otherwise I'll have some steep merge conflicts.
